src/js/components/modals/ServiceFormModal.js

```diff
@@ -0,0 +1,240 @@
+import Ace from 'react-ace';
+import mixin from 'reactjs-mixin';
+import {Modal} from 'reactjs-components';
+import React from 'react';
+import {StoreMixin} from 'mesosphere-shared-reactjs';
+
+import 'brace/mode/json';
+import 'brace/theme/monokai';
+import 'brace/ext/language_tools';
+
+import MarathonStore from '../../stores/MarathonStore';
+import SchemaForm from '../SchemaForm';
+import Service from '../../structs/Service';
+import ServiceUtil from '../../utils/ServiceUtil';
+import ServiceSchema from '../../constants/ServiceSchema';
+import ToggleButton from '../ToggleButton';
+
+const METHODS_TO_BIND = [
+  'getAdvancedSubmit',
+  'handleCancel',
+  'handleClearError',
+  'handleJSONChange',
+  'handleJSONToggle',
+  'handleSubmit',
+  'onMarathonStoreServiceCreateError',
+  'onMarathonStoreServiceCreateSuccess'
+];
+
+class ServiceFormModal extends mixin(StoreMixin) {
+  constructor() {
+    super(...arguments);
+
+    let model =
+      ServiceUtil.createFormModelFromSchema(ServiceSchema);
+    this.state = {
+      app: JSON.stringify({id:'', cmd:''}, null, 2),
+      errorMessage: null,
+      jsonMode: false,
+      model,
+      service: ServiceUtil.createServiceFromFormModel(model)
+    };
+
+    this.store_listeners = [
+      {
+        name: 'marathon',
+        events: ['serviceCreateError', 'serviceCreateSuccess']
+      }
+    ];
+
+    METHODS_TO_BIND.forEach((method) => {
+      this[method] = this[method].bind(this);
+    });
+  }
+
+  componentWillReceiveProps(nextProps) {
+    super.componentWillReceiveProps(...arguments);
+    if (!this.props.open && nextProps.open) {
+      this.resetState();
+    }
+  }
+
+  resetState() {
+    let model = ServiceUtil.createFormModelFromSchema(ServiceSchema);
+    this.setState({
+      app: '',
+      errorMessage: null,
+      jsonMode: false,
+      model,
+      service: ServiceUtil.createServiceFromFormModel(model)
+    });
+  }
+
+  handleClearError() {
+    this.setState({
+      errorMessage: null
+    });
+  }
+
+  handleJSONChange(app) {
+    let {service} = this.state;
+    try {
+      service = new Service(JSON.parse(app));
+    } catch (e) {
+
+    }
+    this.setState({app, service})
+  }
+
+  handleJSONToggle() {
+    let nextState = {};
+    if (!this.state.jsonMode) {
+      let {model} = this.triggerAdvancedSubmit();
```

Comment:
`this.triggerAdvancedSubmit()` is assumed to be existing and defined. However it seems like this is only true if `getAdvancedSubmit` is called first. This seems very brittle and error-prone. 

Beyond hardening this call with some `isFunction` check, I wonder if we should even rely on this get/set mechanism at all.

It seems like we expect the child `SchemaForm` to pass back that function to us so we can store it for later usage ( see https://github.com/dcos/dcos-ui/pull/212/files/90f2cebeb466daabed171e6924ce92aec9cc5b52#diff-076a6fbfc2ab2daac905f988b78234a2R202 ). 
Furthermore we have a lot of references to `Advanced` properties/functions and that smells of tight coupling. 
Can we improve this situation somehow?